snapshotter/devmapper/pool_device.go

```diff
@@ -148,27 +105,71 @@ func (p *PoolDevice) transition(ctx context.Context, deviceName string, tryingSt
 // CreateThinDevice creates new devmapper thin-device with given name and size.
 // Device ID for thin-device will be allocated from metadata store.
 // If allocation successful, device will be activated with /dev/mapper/<deviceName>
-func (p *PoolDevice) CreateThinDevice(ctx context.Context, deviceName string, virtualSizeBytes uint64) error {
+func (p *PoolDevice) CreateThinDevice(ctx context.Context, deviceName string, virtualSizeBytes uint64) (retErr error) {
 	info := &DeviceInfo{
 		Name:  deviceName,
 		Size:  virtualSizeBytes,
 		State: Unknown,
 	}
 
+	var (
+		metadataAdded bool
+		deviceCreated bool
+	)
+
+	defer func() {
```

Comment:
I think this might be more clear?

```go
if err := p.metadata.AddDevice(ctx, info); err != nil {
	return errors.Wrapf(err, "failed to save initial metadata for new thin device %q", deviceName)
}
defer func() {
	if retErr == nil {
		return
	}
	// rollback metadata
	retErr = multierror.Append(result, p.metadata.RemoveDevice(ctx, info.Name)
}()

if err := p.createDevice(ctx, info); err != nil {
	return err
}
defer func() {
	if retErr == nil {
		return
	}
	// rollback creation
	retErr = multierror.Append(result, p.deleteDevice(ctx, info))
}()

return p.activateDevice(ctx, info)
```

This lets the rollback case be coupled with the provisioning case (add and remove are together, create and delete are together) and lets `defer` ordering take care of executing the right code at the right time.  If `p.activateDevice` fails, `p.deleteDevice` will run first followed by `p.metadata.RemoveDevice`.  Similarly, if `p.createDevice` fails, only `p.metadata.RemoveDevice` will run.

The only gross thing here is repeated checking of `retErr` being non-nil, but I'm not sure how to avoid that.

What do you think?

Reply:
I couldn't decide which approach is better, so the suggested one works for me :)